File: Sources/CgRPC/shim/handler.c
/*
 * Copyright 2016, gRPC Authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "internal.h"
#include "cgrpc.h"

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

cgrpc_handler *cgrpc_handler_create_with_server(cgrpc_server *server) {
  cgrpc_handler *handler = (cgrpc_handler *) malloc(sizeof (cgrpc_handler));
  memset(handler, 0, sizeof(cgrpc_handler));
  handler->server = server;
  grpc_metadata_array_init(&(handler->request_metadata_recv));
  grpc_call_details_init(&(handler->call_details));
  handler->completion_queue = grpc_completion_queue_create_for_next(NULL);
  return handler;
}

void cgrpc_handler_destroy(cgrpc_handler *h) {
  grpc_metadata_array_destroy(&(h->request_metadata_recv));
  grpc_call_details_destroy(&(h->call_details));
  if (h->server_call) {
    grpc_call_unref(h->server_call);
  }
  free(h);
}

char *cgrpc_handler_copy_host(cgrpc_handler *h) {
  size_t length = GRPC_SLICE_LENGTH(h->call_details.host);
  char *str = (char *) malloc(length + 1);
  memcpy(str, GRPC_SLICE_START_PTR(h->call_details.host), length);
  str[length] = 0;
  return str;
}

char *cgrpc_handler_copy_method(cgrpc_handler *h) {
  size_t length = GRPC_SLICE_LENGTH(h->call_details.method);
  char *str = (char *) malloc(length + 1);
  memcpy(str, GRPC_SLICE_START_PTR(h->call_details.method), length);
  str[length] = 0;
  return str;
}

char *cgrpc_handler_call_peer(cgrpc_handler *h) {
  return grpc_call_get_peer(h->server_call);
}

cgrpc_call *cgrpc_handler_get_call(cgrpc_handler *h) {
  cgrpc_call *call = (cgrpc_call *) malloc(sizeof(cgrpc_call));
  memset(call, 0, sizeof(cgrpc_call));
  call->call = h->server_call;
  if (call->call) {
    // This retain will be balanced by `cgrpc_call_destroy()`.
    grpc_call_ref(call->call);
  }
  return call;
}

cgrpc_completion_queue *cgrpc_handler_get_completion_queue(cgrpc_handler *h) {
  return h->completion_queue;
}

grpc_call_error cgrpc_handler_request_call(cgrpc_handler *h,
                                           cgrpc_metadata_array *metadata,
                                           long tag) {
  if (h->server_call != NULL) {
    return GRPC_CALL_OK;
  }
  // This fills `h->server_call` with a call with retain count of +1.
  // We'll release that retain in `cgrpc_handler_destroy()`.
  return grpc_server_request_call(h->server->server,
                                  &(h->server_call),
                                  &(h->call_details),
                                  metadata,
                                  h->completion_queue,
                                  h->server->completion_queue,
                                  cgrpc_create_tag(tag));
}

File: Sources/gRPC/Handler.swift
/*
 * Copyright 2016, gRPC Authors All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#if SWIFT_PACKAGE
  import CgRPC
#endif
import Foundation // for String.Encoding

/// A gRPC request handler
public class Handler {
  /// Pointer to underlying C representation
  private var underlyingHandler: UnsafeMutableRawPointer
  
  /// Completion queue for handler response operations
  internal var completionQueue: CompletionQueue
  
  /// Metadata received with the request
  public var requestMetadata: Metadata
  
  /// A Call object that can be used to respond to the request
  internal lazy var call: Call = {
    Call(underlyingCall: cgrpc_handler_get_call(self.underlyingHandler),
         owned: false,
         completionQueue: self.completionQueue)
  }()
  
  /// The host name sent with the request
  public lazy var host: String = {
    if let string = cgrpc_handler_copy_host(self.underlyingHandler) {
      defer {
        cgrpc_free_copied_string(string)
      }
      return String(cString: string, encoding: .utf8)!
    } else {
      return ""
    }
  }()
  
  /// The method name sent with the request
  public lazy var method: String = {
    if let string = cgrpc_handler_copy_method(self.underlyingHandler) {
      defer {
        cgrpc_free_copied_string(string)
      }
      return String(cString: string, encoding: .utf8)!
    } else {
      return ""
    }
  }()
  
  /// The caller address associated with the request
  public lazy var caller: String = {
    String(cString: cgrpc_handler_call_peer(self.underlyingHandler),
           encoding: .utf8)!
  }()
  
  /// Initializes a Handler
  ///
  /// - Parameter underlyingServer: the underlying C representation of the associated server
  init(underlyingServer: UnsafeMutableRawPointer) {
    underlyingHandler = cgrpc_handler_create_with_server(underlyingServer)
    requestMetadata = Metadata()
    completionQueue = CompletionQueue(underlyingCompletionQueue: cgrpc_handler_get_completion_queue(underlyingHandler))
    completionQueue.name = "Handler"
  }
  
  deinit {
    cgrpc_handler_destroy(self.underlyingHandler)
  }
  
  /// Requests a call for the handler
  ///
  /// Fills the handler properties with information about the received request
  ///
  func requestCall(tag: Int) throws {
    let error = cgrpc_handler_request_call(underlyingHandler, requestMetadata.underlyingArray, tag)
    if error != GRPC_CALL_OK {
      throw CallError.callError(grpcCallError: error)
    }
  }
  
  /// Receive the message sent with a call
  ///
  public func receiveMessage(initialMetadata: Metadata,
                             completion: @escaping ((Data?) throws -> Void)) throws {
    let operations = OperationGroup(call: call,
                                    operations: [
                                      .sendInitialMetadata(initialMetadata),
                                      .receiveMessage
    ]) { operationGroup in
      if operationGroup.success {
        try completion(operationGroup.receivedMessage()?.data())
      } else {
        try completion(nil)
      }
    }
    try call.perform(operations)
  }
  
  /// Sends the response to a request
  ///
  /// - Parameter message: the message to send
  /// - Parameter statusCode: status code to send
  /// - Parameter statusMessage: status message to send
  /// - Parameter trailingMetadata: trailing metadata to send
  public func sendResponse(message: Data,
                           statusCode: StatusCode,
                           statusMessage: String,
                           trailingMetadata: Metadata) throws {
    let messageBuffer = ByteBuffer(data: message)
    let operations = OperationGroup(call: call,
                                    operations: [
                                      .receiveCloseOnServer,
                                      .sendStatusFromServer(statusCode, statusMessage, trailingMetadata),
                                      .sendMessage(messageBuffer)
    ]) { operationGroup in
      if operationGroup.success {
        self.shutdown()
      }
    }
    try call.perform(operations)
  }
  
  /// Sends the response to a request
  ///
  /// - Parameter statusCode: status code to send
  /// - Parameter statusMessage: status message to send
  /// - Parameter trailingMetadata: trailing metadata to send
  public func sendResponse(statusCode: StatusCode,
                           statusMessage: String,
                           trailingMetadata: Metadata) throws {
    let operations = OperationGroup(call: call,
                                    operations: [
                                      .receiveCloseOnServer,
                                      .sendStatusFromServer(statusCode, statusMessage, trailingMetadata)
    ]) { operationGroup in
      if operationGroup.success {
        self.shutdown()
      }
    }
    try call.perform(operations)
  }
  
  /// Shuts down the handler's completion queue
  public func shutdown() {
    completionQueue.shutdown()
  }
  
  /// Send initial metadata in response to a connection
  ///
  /// - Parameter initialMetadata: initial metadata to send
  /// - Parameter completion: a completion handler to call after the metadata has been sent
  public func sendMetadata(initialMetadata: Metadata,
                           completion: @escaping (() throws -> Void)) throws {
    let operations = OperationGroup(call: call,
                                    operations: [.sendInitialMetadata(initialMetadata)]) { operationGroup in
      if operationGroup.success {
        try completion()
      } else {
        try completion()
      }
    }
    try call.perform(operations)
  }
  
  /// Receive the message sent with a call
  ///
  /// - Parameter completion: a completion handler to call after the message has been received
  /// - Returns: a tuple containing status codes and a message (if available)
  public func receiveMessage(completion: (@escaping (Data?) throws -> Void)) throws {
    let operations = OperationGroup(call: call, operations: [.receiveMessage]) { operationGroup in
      if operationGroup.success {
        if let message = operationGroup.receivedMessage() {
          try completion(message.data())
        } else {
          try completion(nil)
        }
      } else {
        try completion(nil)
      }
    }
    try call.perform(operations)
  }
  
  /// Sends the response to a request
  ///
  /// - Parameter message: the message to send
  /// - Parameter completion: a completion handler to call after the response has been sent
  public func sendResponse(message: Data,
                           completion: @escaping () throws -> Void) throws {
    let operations = OperationGroup(call: call,
                                    operations: [.sendMessage(ByteBuffer(data: message))]) { operationGroup in
      if operationGroup.success {
        try completion()
      }
    }
    try call.perform(operations)
  }
  
  /// Recognize when the client has closed a request
  ///
  /// - Parameter completion: a completion handler to call after request has been closed
  public func receiveClose(completion: @escaping () throws -> Void) throws {
    let operations = OperationGroup(call: call,
                                    operations: [.receiveCloseOnServer]) { operationGroup in
      if operationGroup.success {
        try completion()
      }
    }
    try call.perform(operations)
  }
  
  /// Send final status to the client
  ///
  /// - Parameter statusCode: status code to send
  /// - Parameter statusMessage: status message to send
  /// - Parameter trailingMetadata: trailing metadata to send
  /// - Parameter completion: a completion handler to call after the status has been sent
  public func sendStatus(statusCode: StatusCode,
                         statusMessage: String,
                         trailingMetadata: Metadata,
                         completion: @escaping (() -> Void)) throws {
    let operations = OperationGroup(call: call,
                                    operations: [
                                      .sendStatusFromServer(statusCode,
                                                            statusMessage,
                                                            trailingMetadata)
    ]) { operationGroup in
      if operationGroup.success {
        completion()
      }
    }
    try call.perform(operations)
  }
}
